File: frontend/lib/api/receipts.ts
import { apiDelete, apiForm, apiGet, apiJson } from "@/lib/api/client";
import type { Receipt, ReceiptInput } from "@/types/receipt";

export function listReceipts(params?: {
  year?: number;
  month?: number;
  category_id?: number;
}): Promise<Receipt[]> {
  const query = new URLSearchParams();
  if (params?.year !== undefined) query.set("year", String(params.year));
  if (params?.month !== undefined) query.set("month", String(params.month));
  if (params?.category_id !== undefined) query.set("category_id", String(params.category_id));
  const qs = query.toString();
  return apiGet<Receipt[]>(`/receipts${qs ? `?${qs}` : ""}`);
}

export function getReceipt(id: number): Promise<Receipt> {
  return apiGet<Receipt>(`/receipts/${id}`);
}

export function createReceipt(payload: ReceiptInput): Promise<Receipt> {
  return apiJson<Receipt>("POST", "/receipts", payload);
}

export function uploadReceipt(payload: ReceiptInput, image: File): Promise<Receipt> {
  const form = new FormData();
  form.append("merchant", payload.merchant);
  form.append("total", String(payload.total));
  form.append("paid_at", payload.paid_at);
  if (payload.memo) form.append("memo", payload.memo);
  if (payload.category_id !== null && payload.category_id !== undefined) {
    form.append("category_id", String(payload.category_id));
  }
  form.append("image", image);
  return apiForm<Receipt>("POST", "/receipts/upload", form);
}

export function updateReceipt(id: number, payload: Partial<ReceiptInput>): Promise<Receipt> {
  return apiJson<Receipt>("PATCH", `/receipts/${id}`, payload);
}

export function deleteReceipt(id: number): Promise<void> {
  return apiDelete(`/receipts/${id}`);
}

File: backend/app/api/routers/receipts.py
from datetime import date
from decimal import Decimal

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.deps import get_db
from app.models.receipt import Receipt
from app.schemas.receipt import ReceiptCreate, ReceiptOCRPreview, ReceiptRead, ReceiptUpdate
from app.services import receipt_service

router = APIRouter(prefix="/receipts", tags=["receipts"])


@router.get("", response_model=list[ReceiptRead])
def list_receipts(
    year: int | None = None,
    month: int | None = None,
    category_id: int | None = None,
    db: Session = Depends(get_db),
) -> list[Receipt]:
    return receipt_service.list_receipts(db, year=year, month=month, category_id=category_id)


@router.get("/{receipt_id}", response_model=ReceiptRead)
def get_receipt(receipt_id: int, db: Session = Depends(get_db)) -> Receipt:
    receipt = db.get(Receipt, receipt_id)
    if receipt is None:
        raise HTTPException(status_code=404, detail="Receipt not found")
    return receipt


@router.post("", response_model=ReceiptRead, status_code=status.HTTP_201_CREATED)
def create_receipt(payload: ReceiptCreate, db: Session = Depends(get_db)) -> Receipt:
    return receipt_service.create_receipt(db, payload)


@router.post("/ocr", response_model=ReceiptOCRPreview)
async def ocr_receipt(image: UploadFile = File(...)) -> ReceiptOCRPreview:
    result = await receipt_service.extract_from_image(image)
    paid_at = result.paid_at.date() if result.paid_at is not None else None
    return ReceiptOCRPreview(
        merchant=result.merchant,
        total=result.total,
        paid_at=paid_at,
        confidence=result.confidence,
    )


@router.post("/upload", response_model=ReceiptRead, status_code=status.HTTP_201_CREATED)
async def upload_receipt(
    merchant: str = Form(...),
    total: Decimal = Form(...),
    paid_at: date = Form(...),
    memo: str | None = Form(None),
    category_id: int | None = Form(None),
    image: UploadFile = File(...),
    db: Session = Depends(get_db),
) -> Receipt:
    payload = ReceiptCreate(
        merchant=merchant,
        total=total,
        paid_at=paid_at,
        memo=memo,
        category_id=category_id,
    )
    return await receipt_service.create_receipt_with_image(db, payload, image)


@router.patch("/{receipt_id}", response_model=ReceiptRead)
def update_receipt(
    receipt_id: int, payload: ReceiptUpdate, db: Session = Depends(get_db)
) -> Receipt:
    receipt = db.get(Receipt, receipt_id)
    if receipt is None:
        raise HTTPException(status_code=404, detail="Receipt not found")
    return receipt_service.update_receipt(db, receipt, payload)


@router.delete("/{receipt_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_receipt(receipt_id: int, db: Session = Depends(get_db)) -> None:
    receipt = db.get(Receipt, receipt_id)
    if receipt is None:
        raise HTTPException(status_code=404, detail="Receipt not found")
    db.delete(receipt)
    db.commit()
